### PR TITLE
results: improve API for recording and printing

### PR DESCRIPTION
**Pull request by matthewrsj**
Add name and description to the Results struct, record all tests in a
Results.Tests slice of Result structs so they can be inspected by the
caller. Print name and description of each failure when printing test
results.

Signed-off-by: Matthew Johnson <matthew.johnson@intel.com>

RFC on the API

### CODE REVIEW

**Comment by matthewrsj** [comment]
@icarus-sparry I updated the API to only have one `Add` function. It does have one boolean in the function signature to tell if it is a  `skipped` test or not. Let me know what you think.

**Comment by matthewrsj** [comment]
ping

File: cmd/pipcheck.go [hunk 1]
```diff
@@ -72,20 +72,15 @@ func runPipCheck(cmd *cobra.Command, args []string) {
 		helpers.Fail(err)
 	}
 
-	if results.Failed > 0 {
-		os.Exit(1)
-	}
+	os.Exit(int(results.Failed))
```
[the remark this answers is not in the document]
Comment (matthewrsj) [reply cmd/pipcheck.go]: Ah, yeah fair enough.

File: internal/testutils/results.go [hunk 2]
```diff
@@ -62,22 +83,32 @@ func (r *Results) PrintJSON(w io.Writer) error {
...
-	r.Total++
-	r.Failures = append(r.Failures, err.Error())
-}
-
-// AddPassed increments the Passed and Total counts
-func (r *Results) AddPassed() {
-	r.Passed++
+// Add increments the appropriate counters according to err and skipped
+// arguments and adds a test the result.Tests field.
+func (r *Results) Add(name, description string, err error, skipped bool) {
```
[the remark this answers is not in the document]
Comment (matthewrsj) [reply internal/testutils/results.go]: How about having the user define a new error for skipped tests?
```go
r.Add("mytest", "run my test", errors.New("skipping bc I wanna"), true)
```

[the remark this answers is not in the document]
Comment (phmccarty) [reply internal/testutils/results.go]: Seems reasonable enough.

**Comment by matthewrsj** [comment]
@phmccarty comments addressed

**Pull request merged**